<launch>

  <arg name="sensor_type" value="bearing" />
  <arg name="sensor_config_file" value="$(find phd_sensor)/config/$(arg sensor_type)_sensor.yaml" />

  <arg name="num_agents" value="1" />
  <arg name="num_targets" value="40" />

  <arg name="base_frame_id" value="base" />
  <arg name="odom_frame_id" value="odom" />
  <arg name="global_frame_id" value="/map" />

  <arg name="rviz_on" value="true" />

  <node name="map" pkg="map_server" type="map_server" output="screen"
        args="$(find maps)/intel/intel.yaml" >
    <param name="frame_id" value="/map" />
  </node>

  <group if="$(arg rviz_on)">
    <node name="rviz" pkg="rviz" type="rviz"
          args="--display-config=$(find phd_filter)/launch/scarab0.rviz" />
  </group>

  <node name="agent_sim" pkg="kinematic_sim" type="kinematic_sim"
        output="screen" respawn="false">
    <param name="publish_freq" value="7"/>
    <param name="pub_global_frame" value="false"/>

    <param name="base_frame_id" value="$(arg base_frame_id)"/>
    <param name="odom_frame_id" value="$(arg odom_frame_id)"/>
    <param name="global_frame_id" value="$(arg global_frame_id)"/>

    <param name="noise_x_sd" value="0.0"/>
    <param name="noise_y_sd" value="0.0"/>
    <param name="noise_th_sd" value="0.0"/>

    <param name="num_agents" value="$(arg num_agents)"/>
    <param name="agent_prefix" value="scarab"/>

    <param name="initial0" value="5.2 5.8 2.16"/>
  </node>

  <node name="target_sim" pkg="kinematic_sim" type="kinematic_sim"
        output="screen" respawn="false">
    <param name="publish_freq" value="7"/>
    <param name="pub_global_frame" value="true"/>

    <param name="base_frame_id" value="$(arg base_frame_id)"/>
    <param name="odom_frame_id" value="$(arg odom_frame_id)"/>
    <param name="global_frame_id" value="$(arg global_frame_id)"/>

    <param name="noise_x_sd" value="0.0"/>
    <param name="noise_y_sd" value="0.0"/>
    <param name="noise_th_sd" value="0.0"/>

    <param name="num_agents" value="$(arg num_targets)"/>
    <param name="agent_prefix" value="target"/>

    <rosparam file="$(find phd_filter)/launch/targets_intel40.yaml" command="load" />
  </node>

  <node name="target_aggregator" pkg="laser_simulator" type="pose_aggregator"
        output="screen" respawn="false">
    <remap from="pose_array" to="target_array"/>

    <param name="publish_freq" value="7"/>

    <param name="num_agents" value="$(arg num_targets)"/>
    <param name="agent_prefix" value="target"/>
    <param name="frame_id" value="base"/>
    <param name="pose_topic" value="gt_pose"/>
  </node>

  <group ns="scarab0">
    <node name="laser" pkg="laser_simulator" type="laser_simulator"
          output="screen" respawn="false">
      <param name="frame_id" value="/scarab0/$(arg base_frame_id)"/>
      <param name="offset/x" value="0.0"/>
      <param name="offset/y" value="0.0"/>
      <param name="offset/z" value="0.2"/>

      <param name="noise_sd" value="0.003"/>

      <rosparam file="$(find laser_simulator)/models/UTM_30.yaml" command="load" />
      <rosparam file="$(find laser_simulator)/config/models.yaml" command="load" />

      <remap from="~map" to="/map"/>
      <remap from="~odom" to="/scarab0/gt_odom"/>
      <remap from="~pose_array" to="/pose_array"/>
      <remap from="~scan" to="/scarab0/scan" />
    </node>

    <node name="$(anon transformer)" pkg="tf" type="static_transform_publisher"
          args="0.0 0.0 0.0 0.0 0.0 0.0 $(arg global_frame_id)
                /scarab0/$(arg odom_frame_id) 100" />

    <node name="sensor_sim" pkg="phd_sensor" type="$(arg sensor_type)_sim"
          output="screen" respawn="false" launch-prefix="">
      <remap from="target_array" to="/target_aggregator/pose_array"/>
      <remap from="pose" to="gt_pose" />
      <remap from="measurements" to="measurements_raw"/>
      <remap from="map" to="/map"/>

      <param name="publish_freq" value="5.0"/>

      <param name="frame_id" value="base"/>

      <rosparam file="$(arg sensor_config_file)" command="load" />
    </node>

    <node name="sensor_viz" pkg="phd_sensor" type="$(arg sensor_type)_vis"
          output="screen" respawn="false">
      <param name="frame_id" value="base"/>
    </node>

    <node name="sensor_throttle" pkg="phd_sensor" type="$(arg sensor_type)_throttle"
          output="screen" respawn="false">
      <remap from="measurements" to="measurements_raw"/>
      <remap from="measurements_throttle" to="measurements"/>

      <param name="min_time" value="0.1"/>
      <param name="max_time" value="1.0"/>

      <param name="min_angle" value="0.785"/> <!-- pi/4 radians -->
      <param name="min_step" value="2.0"/>
    </node>

    <node name="goal_to_action" pkg="hfn" type="goal_to_action.py" />

    <node name="navigation" pkg="hfn" type="hfn" respawn="false"
          output="screen" launch-prefix="" >
      <remap from="pose" to="gt_pose"/>
      <remap from="odom" to="gt_odom"/>
      <remap from="map" to="/map"/>

      <param name="base_frame_id" value="/scarab0/$(arg base_frame_id)" />
      <param name="map_frame_id" value="$(arg global_frame_id)" />

      <param name="robot_radius" value="0.23" />
      <param name="safety_margin" value="0.0" />

      <param name="v_opt" value="0.45" />
      <param name="w_max" value="0.65" />
      <param name="waypoint_thresh" value="0.2" />
      <param name="waypoint_spacing" value="0.4" />

      <param name="cost_occ_prob" value="0.0" />
      <param name="cost_occ_dist" value="0.0" />
      <param name="max_occ_dist" value="1.0" />
      <param name="lethal_occ_dist" value="0.2" />

      <param name="alpha_thresh" value="2.094" />
      <param name="path_margin" value="0.5" />
      <param name="allow_unknown_path" value="false" />

      <param name="goal_tolerance" value="0.5" />
      <param name="timeout" value="120" />
    </node>

    <node name="phd_filter" pkg="phd_filter" type="phd_filter_$(arg sensor_type)_node"
          output="screen" respawn="false" launch-prefix="">
      <remap from="map" to="/map" />
      <remap from="pose" to="gt_pose" />

      <param name="num_targets" value="10.0" />
      <param name="max_value" value="0.01" />
      <param name="width" value="30.0" />
      <param name="height" value="30.0" />

      <rosparam file="$(arg sensor_config_file)" ns="sensor" command="load" />
      <rosparam file="$(find phd_filter)/config/static_motion_model.yaml" ns="motion" command="load" />
    </node>
  </group>


</launch>
